File: backend/static/js/panels/health.js
/* Knowing what is actually up.
 *
 * The grid reports every camera as live, including the ones that are not. So
 * this panel puts the reported number and the measured number side by side, and
 * the gap between them is the whole point: it is the count of cameras nobody has
 * verified.
 *
 * Probing follows the video wall rather than sweeping the estate. This grid
 * permits roughly one session per address and measurably degrades under
 * parallel opens -- at ten concurrent opens, five cameras returned no frames at
 * all, while serially four of them recovered. A "check everything" button would
 * therefore make the estate look worse than it is, so there is not one.
 */

import { api, session } from '../api.js';
import { state, visibleCameras } from '../store.js';
import { el, fill, icon, fmtNum, fmtTime, fmtDuration, toast, into } from '../ui.js';
import * as mapView from '../map.js';

export function render(host) {
  into(host, async () => {
    const { summary, results } = await api.health();
    const byId = new Map(results.map((r) => [r.camera_id, r]));

    return [
      el('div.panel-section', {},
        el('div.metrics', {},
          metric('Reported up', summary.reported_online, 'by the grid', ''),
          metric('Confirmed', summary.confirmed_online, 'opened and decoded', 'ok'),
          metric('Unverified', summary.unverified, 'never probed',
            summary.unverified ? 'warn' : ''),
          metric('Degraded', summary.degraded, 'up but unusable',
            summary.degraded ? 'warn' : ''),
          metric('Failed', summary.failed, 'did not open',
            summary.failed ? 'crit' : ''),
        ),
        el('div.note', { style: { marginTop: '12px' },
          text: `${fmtNum(summary.reported_online)} cameras report themselves online. `
              + `${fmtNum(summary.confirmed_online)} have actually been opened and decoded `
              + `by this process. The rest is an assumption, not a measurement.` }),
      ),

      session.canAct ? probeControls(host) : null,

      el('div.panel-section', {},
        el('div.label', { style: { marginBottom: '10px' }, text: 'Per camera' }),
        ...visibleCameras().map((cam) => cameraRow(cam, byId.get(cam.camera_id))),
      ),
    ];
  });
}

function metric(label, value, sub, kind) {
  return el(`div.metric${kind ? '.' + kind : ''}`, {},
    el('div.m-k', { text: label }),
    el('div.m-v', { text: fmtNum(value) }),
    el('div.m-s', { text: sub }));
}

function probeControls(host) {
  const targets = () => (state.wall.length ? state.wall
    : visibleCameras().slice(0, 6).map((c) => c.camera_id));

  return el('div.panel-section', {},
    el('div.label', { style: { marginBottom: '8px' }, text: 'Measure' }),
    el('div', { style: { display: 'flex', gap: '8px', flexWrap: 'wrap' } },
      el('button.btn.primary', {
        onclick: async (event) => {
          const button = event.currentTarget;
          const ids = targets();
          if (!ids.length) { toast('Nothing to probe', 'Add cameras to the wall first.', 'warn'); return; }
          button.disabled = true;
          fill(button, 'Probing…');
          try {
            const result = await api.probe(ids, true);
            const up = result.results.filter((r) => r.status === 'ONLINE').length;
            toast('Probe complete',
              `${up} of ${result.results.length} confirmed online.`,
              up === result.results.length ? 'ok' : 'warn');
            render(host);
          } catch (error) {
            toast('Probe failed', error.message, 'crit');
            button.disabled = false;
            fill(button, icon('health', 13), 'Probe');
          }
        },
      }, icon('health', 13),
         state.wall.length ? `Probe the ${state.wall.length} on the wall` : 'Probe 6 cameras'),
    ),
    el('div.faint', { style: { fontSize: '10px', marginTop: '9px', lineHeight: '1.55' },
      text: 'Probing opens a real stream per camera and blocks for up to 25 s each. '
          + 'It is serialised behind the same concurrency cap the ingest workers '
          + 'use, because this grid returns no frames at all under parallel opens.' }),
  );
}

function cameraRow(cam, probe) {
  const measured = probe?.status;
  const disagrees = measured && measured !== cam.status;

  return el('button.row', {
    onclick: () => mapView.selectCamera(cam.camera_id),
    title: probe?.note || 'Never probed',
  },
    el('div.row-main', {},
      el('div.row-title', { text: cam.location_name || cam.camera_id }),
      el('div.row-sub', {},
        el('span.mono', { text: cam.camera_id }),
        probe ? ` · probed ${fmtTime(probe.checked_at)}` : ' · never probed'),
    ),
    el('div.row-side', { style: { display: 'flex', gap: '6px', alignItems: 'center' } },
      el('span.tag', { text: cam.status, title: 'Reported by the grid' }),
      measured
        ? el('span.tag.' + tagClass(measured), { text: measured, title: 'Measured here' })
        : el('span.tag', { style: { opacity: '0.5' }, text: '—' }),
      probe?.open_latency_s !== null && probe?.open_latency_s !== undefined
        ? el('span.faint.mono', { style: { fontSize: '10px', minWidth: '42px' },
            text: fmtDuration(probe.open_latency_s) })
        : null,
    ),
    /* A disagreement between the two columns is the finding, so it is marked
     * rather than left for the reader to spot. */
    disagrees ? el('span.dot', { style: { color: 'var(--warn)' },
      title: `Grid says ${cam.status}, measurement says ${measured}` }) : null,
  );
}

const tagClass = (status) => ({ ONLINE: 'ok', DEGRADED: 'warn', OFFLINE: 'crit' }[status] || '');

File: backend/app/services/health.py
"""Measured camera availability.

The grid reports every camera as live, including the ones that are not. So
availability is measured here rather than trusted: a camera is ONLINE because
this process opened it and decoded a frame, not because a catalogue said so.

Three outcomes, and the middle one is the interesting one:

  * **ONLINE**   -- opened, and decoded a frame with real content.
  * **DEGRADED** -- opened, but the frames are unusable. On this grid that is
    mostly CAM-12's H.265 smear: the capture succeeds, frames arrive, and every
    one of them is a grey mush. A camera that reports itself up and delivers
    nothing readable is the single most misleading state on the estate, so it
    gets its own status rather than being rounded to ONLINE.
  * **OFFLINE**  -- did not open inside the timeout.

Probing policy
--------------
Probing follows the video wall rather than sweeping the estate. This grid
permits roughly one session per address and measurably degrades under parallel
opens -- at ten concurrent opens, five cameras returned no frames at all, while
serially four of them recovered. So probes are serialised behind the same global
stream semaphore the ingest workers use, and a camera probed recently is not
re-probed.
"""
from __future__ import annotations

import logging
import threading
import time
from dataclasses import dataclass, field
from datetime import datetime, timezone

import cv2
import numpy as np

from ..config import settings
from ..models import Camera, CameraStatus
from ..pipeline.worker import StreamWorker

log = logging.getLogger("sentinel.health")

#: How long a probe result stays fresh. A camera checked inside this window is
#: reported from cache rather than reopened -- reopening costs a stream slot and
#: this grid does not tolerate churn.
FRESH_FOR_S = 300.0

#: Frames to decode before judging a camera. One frame is not enough: the first
#: frame after a join is frequently a partially-decoded reference frame that
#: looks like smear on a perfectly healthy camera.
PROBE_FRAMES = 3

#: Opens on this grid have been measured anywhere from 1.8 s to 275 s. A probe
#: cannot wait that long, so a slow-but-alive camera can come back OFFLINE here.
#: That is stated in the result rather than hidden.
PROBE_TIMEOUT_S = 25.0

#: Below this standard deviation a frame carries no structure -- a black frame,
#: a dropped-signal blue screen, or the grey mush CAM-12 produces. Measured: a
#: healthy night frame on this grid reads 28-60, CAM-12's smear reads under 6.
FLAT_FRAME_STD = 8.0


@dataclass
class ProbeResult:
    camera_id: str
    status: str
    open_latency_s: float | None = None
    frames_decoded: int = 0
    frame_std: float | None = None
    note: str = ""
    checked_at: datetime = field(default_factory=lambda: datetime.now(timezone.utc))

    def to_dict(self) -> dict:
        return {
            "camera_id": self.camera_id,
            "status": self.status,
            "open_latency_s": (round(self.open_latency_s, 2)
                               if self.open_latency_s is not None else None),
            "frames_decoded": self.frames_decoded,
            "frame_std": round(self.frame_std, 1) if self.frame_std is not None else None,
            "note": self.note,
            "checked_at": self.checked_at.isoformat(),
        }


_results: dict[str, ProbeResult] = {}
_lock = threading.Lock()


def cached(camera_id: str) -> ProbeResult | None:
    """The last probe of this camera, if it is still fresh."""
    with _lock:
        result = _results.get(camera_id)
    if result is None:
        return None
    age = (datetime.now(timezone.utc) - result.checked_at).total_seconds()
    return result if age < FRESH_FOR_S else None


def all_results() -> list[dict]:
    with _lock:
        return [r.to_dict() for r in
                sorted(_results.values(), key=lambda r: r.camera_id)]


def probe(camera: Camera, *, force: bool = False) -> ProbeResult:
    """Open one camera, decode a few frames, and judge it.

    Blocking, and it holds a stream slot for the duration -- call it off the
    request thread.
    """
    if not force:
        fresh = cached(camera.camera_id)
        if fresh is not None:
            return fresh

    if not camera.stream_url:
        return _record(ProbeResult(camera_id=camera.camera_id, status="OFFLINE",
                                   note="no stream URL in the registry"))

    worker = StreamWorker(
        camera.stream_url,
        camera_id=camera.camera_id,
        fallback_url=camera.hls_url,
        sample_interval_ms=0,
        open_timeout_s=PROBE_TIMEOUT_S,
        # One shot. A probe that retries is no longer measuring availability,
        # it is measuring patience.
        backoff_base=PROBE_TIMEOUT_S,
        backoff_cap=PROBE_TIMEOUT_S,
        read_failure_grace=8,
    )

    started = time.time()
    frames: list[np.ndarray] = []
    deadline = started + PROBE_TIMEOUT_S

    try:
        for event in worker.stream():
            frames.append(event.frame)
            if len(frames) >= PROBE_FRAMES or time.time() > deadline:
                break
    except Exception as exc:
        log.warning("%s: probe failed: %s", camera.camera_id, exc)
    finally:
        worker.stop()

    latency = time.time() - started

    if not frames:
        return _record(ProbeResult(
            camera_id=camera.camera_id, status="OFFLINE", open_latency_s=latency,
            note=f"no frame decoded within {PROBE_TIMEOUT_S:.0f}s. Opens on this "
                 f"grid have been measured up to 275s, so a slow camera can "
                 f"read as offline here."))

    # Judge on the last frame: the first after a join is often a partially
    # decoded reference frame that looks like smear on a healthy camera.
    sharpest = max(float(np.asarray(f).std()) for f in frames)

    if sharpest < FLAT_FRAME_STD:
        return _record(ProbeResult(
            camera_id=camera.camera_id, status="DEGRADED", open_latency_s=latency,
            frames_decoded=len(frames), frame_std=sharpest,
            note="opens and delivers frames, but they carry no structure "
                 f"(std {sharpest:.1f} < {FLAT_FRAME_STD}). Reports itself up "
                 "while being unusable."))

    return _record(ProbeResult(
        camera_id=camera.camera_id, status="ONLINE", open_latency_s=latency,
        frames_decoded=len(frames), frame_std=sharpest,
        note=f"decoded {len(frames)} frames in {latency:.1f}s"))


def _record(result: ProbeResult) -> ProbeResult:
    with _lock:
        _results[result.camera_id] = result
    return result


def apply_to_registry(db, result: ProbeResult) -> None:
    """Write a probe result back onto the camera row.

    Kept separate from ``probe`` so the probe itself needs no database session
    and can run on a worker thread without holding one open for 25 seconds.
    """
    from . import registry

    camera = registry.get_camera(db, result.camera_id)
    if camera is None:
        return
    camera.status = CameraStatus(result.status)
    camera.open_latency_s = result.open_latency_s
    camera.health_note = result.note
    if result.status == "ONLINE":
        camera.last_seen = result.checked_at
    db.commit()


def summary(cameras: list[Camera]) -> dict:
    """Registry-reported vs measured availability.

    The gap between the two columns is the whole point of this panel: it is the
    number of cameras the grid claims are up that nobody has actually verified.
    """
    with _lock:
        measured = dict(_results)

    reported_online = sum(c.status == CameraStatus.ONLINE for c in cameras)
    checked = [c for c in cameras if c.camera_id in measured]

    confirmed = sum(measured[c.camera_id].status == "ONLINE" for c in checked)
    degraded = sum(measured[c.camera_id].status == "DEGRADED" for c in checked)
    failed = sum(measured[c.camera_id].status == "OFFLINE" for c in checked)

    return {
        "total": len(cameras),
        "reported_online": reported_online,
        "probed": len(checked),
        "unverified": len(cameras) - len(checked),
        "confirmed_online": confirmed,
        "degraded": degraded,
        "failed": failed,
        "fresh_for_s": FRESH_FOR_S,
    }
